### ISSUE TITLE
'rustc' panicked at 'attempted .var_id() on invalid Static(..)'

### ISSUE THREAD

**Issue by kaksmet** [Manishearth/rust-clippy#746]
Running clippy on [jpeg-decoder](https://github.com/kaksmet/jpeg-decoder) causes an ICE.
I am using today's nightly rust (2016-03-07) and the latest published clippy.

```
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
thread 'rustc' panicked at 'attempted .var_id() on invalid Static(DefId { krate: 0, node: DefIndex(27) => decoder::UNZIGZAG }, false)', ../src/librustc/middle/def.rs:119
stack backtrace:
   1:        0x10f33fbd8 - sys::backtrace::tracing::imp::write::hdcc91e23d52f1e9feuv
   2:        0x10f348585 - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.44400
   3:        0x10f348087 - panicking::default_handler::h5bfcd298419b6ef54Wz
   4:        0x10f30a3f6 - sys_common::unwind::begin_unwind_inner::hbee05b49f1a706c20su
   5:        0x10f30aeee - sys_common::unwind::begin_unwind_fmt::habb588106280858f6ru
   6:        0x10bcf644a - middle::def::Def::var_id::h9421048a3aec9f7cNws
   7:        0x112920ec3 - loops::VarVisitor<'v, 't>.Visitor<'v>::visit_expr::hcd574d0df5a387e2Gkg
   8:        0x11291fb9c - intravisit::walk_expr::h10051942649284778346
   9:        0x112921039 - loops::VarVisitor<'v, 't>.Visitor<'v>::visit_expr::hcd574d0df5a387e2Gkg
  10:        0x11291f61e - intravisit::walk_expr::h10051942649284778346
  11:        0x112921039 - loops::VarVisitor<'v, 't>.Visitor<'v>::visit_expr::hcd574d0df5a387e2Gkg
  12:        0x11291fb9c - intravisit::walk_expr::h10051942649284778346
  13:        0x112921039 - loops::VarVisitor<'v, 't>.Visitor<'v>::visit_expr::hcd574d0df5a387e2Gkg
  14:        0x11291fa1c - intravisit::walk_expr::h10051942649284778346
  15:        0x112921039 - loops::VarVisitor<'v, 't>.Visitor<'v>::visit_expr::hcd574d0df5a387e2Gkg
  16:        0x112923410 - intravisit::walk_stmt::h2186350535004613219
  17:        0x11292336c - intravisit::Visitor::visit_stmt::h8788178206728062161
  18:        0x112923261 - intravisit::walk_block::h16959476525822426093
  19:        0x11292312c - intravisit::Visitor::visit_block::h4525436756485254811
  20:        0x11291f9af - intravisit::walk_expr::h10051942649284778346
  21:        0x11291b26a - loops::check_for_loop_range::h43ac53f4a7e48d3fBOf
  22:        0x11291a224 - loops::check_for_loop::h2aeba32ebd234adbENf
  23:        0x112820678 - loops::LoopsPass.LateLintPass::check_expr::h051afc1cd0faa03eBGf
  24:        0x10bf140d9 - lint::context::LintContext::with_lint_attrs::h17792434420503030058
  25:        0x10bf1994f - lint::context::LintContext::with_lint_attrs::h11630826001916547603
  26:        0x10bf15bc4 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_decl::h443372e119582f69Giv
  27:        0x10bf14eee - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_block::h1a8d2bf23bf7167aKgv
  28:        0x10bf14bce - intravisit::walk_expr::h2493906109654503125
  29:        0x10bf14165 - lint::context::LintContext::with_lint_attrs::h17792434420503030058
  30:        0x10bf15900 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_stmt::h9370a6fbd09f78b1D6u
  31:        0x10bf14eee - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_block::h1a8d2bf23bf7167aKgv
  32:        0x10bf14bce - intravisit::walk_expr::h2493906109654503125
  33:        0x10bf14165 - lint::context::LintContext::with_lint_attrs::h17792434420503030058
  34:        0x10bf153f9 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_arm::h4fcf354091fb0679Xhv
  35:        0x10bf1497e - intravisit::walk_expr::h2493906109654503125
  36:        0x10bf14165 - lint::context::LintContext::with_lint_attrs::h17792434420503030058
  37:        0x10bf14f19 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_block::h1a8d2bf23bf7167aKgv
  38:        0x10bf14bce - intravisit::walk_expr::h2493906109654503125
  39:        0x10bf14165 - lint::context::LintContext::with_lint_attrs::h17792434420503030058
  40:        0x10bf153f9 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_arm::h4fcf354091fb0679Xhv
  41:        0x10bf1497e - intravisit::walk_expr::h2493906109654503125
  42:        0x10bf14165 - lint::context::LintContext::with_lint_attrs::h17792434420503030058
  43:        0x10bf14f19 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_block::h1a8d2bf23bf7167aKgv
  44:        0x10bf148ec - intravisit::walk_expr::h2493906109654503125
  45:        0x10bf14165 - lint::context::LintContext::with_lint_attrs::h17792434420503030058
  46:        0x10bf153f9 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_arm::h4fcf354091fb0679Xhv
  47:        0x10bf1497e - intravisit::walk_expr::h2493906109654503125
  48:        0x10bf14165 - lint::context::LintContext::with_lint_attrs::h17792434420503030058
  49:        0x10bf1994f - lint::context::LintContext::with_lint_attrs::h11630826001916547603
  50:        0x10bf15bc4 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_decl::h443372e119582f69Giv
  51:        0x10bf14eee - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_block::h1a8d2bf23bf7167aKgv
  52:        0x10bf14bce - intravisit::walk_expr::h2493906109654503125
  53:        0x10bf14165 - lint::context::LintContext::with_lint_attrs::h17792434420503030058
  54:        0x10bf14f19 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_block::h1a8d2bf23bf7167aKgv
  55:        0x10bf14bce - intravisit::walk_expr::h2493906109654503125
  56:        0x10bf14165 - lint::context::LintContext::with_lint_attrs::h17792434420503030058
  57:        0x10bf153f9 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_arm::h4fcf354091fb0679Xhv
  58:        0x10bf1497e - intravisit::walk_expr::h2493906109654503125
  59:        0x10bf14165 - lint::context::LintContext::with_lint_attrs::h17792434420503030058
  60:        0x10bf15900 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_stmt::h9370a6fbd09f78b1D6u
  61:        0x10bf14eee - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_block::h1a8d2bf23bf7167aKgv
  62:        0x10bf148ec - intravisit::walk_expr::h2493906109654503125
  63:        0x10bf14165 - lint::context::LintContext::with_lint_attrs::h17792434420503030058
  64:        0x10bf15900 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_stmt::h9370a6fbd09f78b1D6u
  65:        0x10bf14eee - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_block::h1a8d2bf23bf7167aKgv
  66:        0x10bf0f5ab - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_fn::hf7272ff0c3c4d33bm7u
  67:        0x10bf1cd4c - intravisit::walk_impl_item::h1304523565438941419
  68:        0x10bf1c5c7 - lint::context::LintContext::with_lint_attrs::h16203647713301038855
  69:        0x10bf0ae69 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_item::h99877c0635d1b14901u
  70:        0x10bf0fb3c - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_mod::hdd06b82b88559662sev
  71:        0x10bf0aea5 - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_item::h99877c0635d1b14901u
  72:        0x10bf0fb3c - lint::context::LateContext<'a, 'tcx>.hir_visit..Visitor<'v>::visit_mod::hdd06b82b88559662sev
  73:        0x10bf329ad - lint::context::check_crate::hffa5097d0cc92af363v
  74:        0x10ae6f90b - driver::phase_3_run_analysis_passes::_$u7b$$u7b$closure$u7d$$u7d$::closure.29847
  75:        0x10ae69ee9 - middle::ty::context::TyCtxt<'tcx>::create_and_enter::h490329377699204806
  76:        0x10ae6696e - driver::phase_3_run_analysis_passes::h3048820087593366542
  77:        0x10ae392c8 - driver::compile_input::h832b24e39ace76b6Kca
  78:        0x10ae2702b - run_compiler::hc9c072fd47d44e4f8Oc
  79:        0x10ae24682 - sys_common::unwind::try::try_fn::h6466067478841458575
  80:        0x10f33d5ab - __rust_try
  81:        0x10f3355b3 - sys_common::unwind::inner_try::hd157bd7fdf16dec72pu
  82:        0x10ae24f19 - boxed::F.FnBox<A>::call_box::h11501500639451413514
  83:        0x10f34757c - sys::thread::Thread::new::thread_start::h5b180fca632674e3vbz
  84:     0x7fff8cb9d898 - _pthread_body
  85:     0x7fff8cb9d729 - _pthread_start

Could not compile `jpeg-decoder`.
```

**Comment by mcarton** [comment]
Minimal example reproducing the bug:
```rust
#![feature(plugin)]
#![plugin(clippy)]
#![deny(clippy)]

static UNZIGZAG: [usize; 4] = [ 0,  1,  8, 16 ];

fn main() {
    for j in 0..4 {
        UNZIGZAG[j];
    }
}
```

Looks like we have a problem with statics.

**Comment by mcarton** [comment]
@kaksmet Fixed in 0.0.48, thanks for the report
